caffe2/core/storage.h

```diff
@@ -77,117 +75,150 @@ class CAFFE2_API StorageImpl : public c10::intrusive_ptr_target {
     return data_ptr_.get();
   }
 
-  DataPtr& data_ptr() {
+  at::DataPtr& data_ptr() {
     return data_ptr_;
   }
 
-  const DataPtr& data_ptr() const {
+  const at::DataPtr& data_ptr() const {
     return data_ptr_;
   }
 
-  void set_dtype(const DataType& data_type) {
+  // Returns the previous data_ptr
+  at::DataPtr set_data_ptr(at::DataPtr&& data_ptr) {
+    std::swap(data_ptr_, data_ptr);
+    return std::move(data_ptr);
+  };
+
+  void set_dtype(const TypeMeta& data_type) {
+    int64_t capacity = numel_ * data_type_.itemsize();
     data_type_ = data_type;
+    set_capacity(capacity);
   }
 
-  const DataType& dtype() const {
+  const TypeMeta& dtype() const {
     return data_type_;
   }
 
+  const at::Allocator* allocator() const {
+    return allocator_;
+  };
+  // You generally shouldn't use this method, but it is occasionally
+  // useful if you want to override how a tensor will be reallocated,
+  // after it was already allocated (and its initial allocator was
+  // set)
+  void set_allocator(at::Allocator* allocator) {
+    allocator_ = allocator;
+  }
+
   size_t capacity() const {
-    return capacity_;
+    return numel_ * itemsize();
   }
 
   int64_t numel() const {
-    return capacity_ / itemsize();
+    return numel_;
   }
 
   // TODO: remove later
   void set_numel(int64_t numel) {
-    capacity_ = numel * itemsize();
+    numel_ = numel;
+  }
+
+  // TODO: remove later
+  void set_capacity(int64_t capacity) {
+    if (capacity) {
+      // CAFFE_ENFORCE(
+      //     capacity % data_type_.itemsize() == 0,
+      //     "capacity " + caffe2::to_string(capacity) +
+      //         " is not divisible by itemsize " +
+      //         caffe2::to_string(data_type_.itemsize()) + " of type name " +
+      //         data_type_.name() + ".");
+      numel_ = capacity / data_type_.itemsize();
```

Comment:
It might be (marginally) safer to round up rather than down here, because in that case you will be guaranteed to provide enough bytes to store whatever was previously requested.

Reply:
(This is of course assuming that you don't have the `CAFFE_ENFORCE`. I'm not saying you have to do this, but if you suspect there's some off by one error on the boundary, I'd definitely make this change.)